### PR TITLE
Feat pt: Support multi structures input for single frame

### PR DESCRIPTION
**Pull request by Chengqian-Zhang**
Support multi structures input for one frame:

background:
When we predict the properties of small molecules, sometimes multiple 3D conformations are generated from the SMILES format structure, but their targets are the same. In each epoch we would like to pick only one of the 3D configurations as input.

Impletation:
When the file "multistru" is present in the data folder (like nopbc), "coord.npy" can be written in the format (nframe * struc_num * natom * 3), where *struc_num* is the number of 3D conformations for each structure.

### CODE REVIEW

File: deepmd/utils/data.py [hunk 1]
```diff
@@ -269,6 +271,11 @@ def get_item_torch(self, index: int) -> dict:
         """
         i = bisect.bisect_right(self.prefix_sum, index)
         frames = self._load_set(self.dirs[i])
+        if self.multistru:
+            size = frames["coord"].shape[1]
+            rng = np.random.default_rng()
+            sample_idx = rng.integers(low=0, high=size, size=1)[0]
```
Comment (Chengqian-Zhang) [inline deepmd/utils/data.py:277]: Is it really random?

File: deepmd/utils/data.py
Comment (njzjz) [inline deepmd/utils/data.py:1]: Shouldn't it be documented? Otherwise no one knows how to use it.

File: deepmd/utils/data.py (hunk 1 above)
Comment (njzjz) [reply deepmd/utils/data.py:277]: We use the random generator in `deepmd.utils.rndom`. The generator here cannot be controlled.

File: deepmd/utils/data.py [hunk 2]
```diff
@@ -621,43 +633,49 @@ def _load_data(
...
-                                )
-                                data = tmp
-                            else:
-                                natoms = natoms_sel
-                                idx_map = idx_map_sel
-                                ndof = ndof_ * natoms
-                        elif data.size == nframes * natoms * ndof_:
-                            if output_natoms_for_type_sel:
-                                pass
+                if key == "coord" and self.multistru:
```
Comment (njzjz) [inline deepmd/utils/data.py:636]: Do we assume aparam and other properties are the same for all conformations?

**Review by njzjz — CHANGES_REQUESTED** [review]

File: deepmd/utils/data.py (hunk 1 above)
Comment (Chengqian-Zhang) [reply deepmd/utils/data.py]: I have change it to 
`from deepmd.utils import random as dp_random`
`sample_idx = dp_random.choice(range(num_stru), size=1)[0]`
But I find that when num_stru = 11, the `sample_idx` is 1,4,8,10 all the time. I hope that the value is randomly selected from `range(0--10)`. How can I solve it?

File: deepmd/utils/data.py
Comment (Chengqian-Zhang) [reply deepmd/utils/data.py:1]: I add doc to `deepmd/doc`. Should I add doc to other places?

File: deepmd/utils/data.py (hunk 2 above)
Comment (Chengqian-Zhang) [reply deepmd/utils/data.py:635]: I haven't thought of a situation where there are different other properties(aparam fparam) of all conformations. The reason I add this feature is that when one only have molecules in SMILES format, one may generate several possible conformations using tools such as rdkit. Do you have more better ideas?

**Review by wanghan-iapcm — COMMENTED** [review]
I do not find the the feature is necessary. if different structures have different labels, they should not be placed in one dataset, because the dp model cannot fit  to diverging labels.

**Comment by Chengqian-Zhang** [comment]
@wanghan-iapcm  Different 3D conformations of a molecule have exactly the same label, and each epoch takes one of all 3D conformations of this molecule.

**Comment by wanghan-iapcm** [comment]
> @wanghan-iapcm Different 3D conformations of a molecule have exactly the same label, and each epoch takes one of all 3D conformations of this molecule. Because molecule may only has SMILES, the exact 3D sturcture is unknown.

I do not find any reason why different 3D configurations cannot be place in one batch.

**Comment by Chengqian-Zhang** [comment]
There is currently no need for this PR, I will close it.

**Pull request closed without merging**